### PR TITLE
ROX-16667: Optional encryption key secret mount to Helm and Operator

### PR DESCRIPTION
**Pull request by vladbologa**
## Description

This PR adds support in the Helm charts and the Operator for an option to encrypt notifier secrets.

This is done by:
* creating a secret called `central-encryption-key` in the Central namespace, and then:
* setting `central.encryptNotifierSecrets` to `true` in the Helm chart, or
* setting `spec.central.notifierSecretsEncryptionEnabled` to `true` in the Central CR

This will in turn
* export `ROX_ENCRYPT_NOTIFIER_SECRETS="true"` in the Central container
* mount the encryption key at `/run/secrets/stackrox.io/central-encryption-key/`

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

As the feature is currently not intended to be used directly by users, I did not document the changes in the Helm charts, and I did not create a documentation PR. 

### Here I tell how I validated my change

1. installed Central in a local colima cluster, using the operator
2. ran `/bin/bash` in the Central container, checked that `ROX_ENCRYPT_NOTIFIER_SECRETS` is not defined
3. created a secret called `central-encryption-key` and applied it to the Central namespace
4. patched to Central CR and added `spec.central.notifierSecretsEncryption.enabled: true`
5. verified that the Operator reconciled the Central and restarted the Central pod
6. ran `bin/bash` in the Central container again, this time making sure that the env var `ROX_ENCRYPT_NOTIFIER_SECRETS` exists and is set to `true`, and that the contents of the secret is available at `/run/secrets/stackrox.io/central-encryption-key/`

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by porridge** [comment]
I took a look and it seems that the operator e2e tests failed due to scanner initialization slowness.

At least in the `basic-sc` case, it was doing something...  for 9 minutes (!) between finishing DB migrations and starting to load NVD definitions.

```
[pod/scanner-697f44578b-72w8j/scanner] {"Event":"database migration ran successfully","Level":"info","Location":"pgsql.go:205","Time":"2023-09-26 18:20:51.285129"}
[pod/scanner-697f44578b-72w8j/scanner] {"Event":"Loading NVD definitions into cache","Level":"info","Location":"singleton.go:24","Time":"2023-09-26 18:29:18.650833"}
[pod/scanner-697f44578b-72w8j/scanner] {"Event":"Loading definitions directory","Level":"info","Location":"load.go:19","Time":"2023-09-26 18:29:18.651079","dir":"/nvd_definitions"}
```

**Comment by porridge** [comment]
On a retry, [this](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stackrox_stackrox/7889/pull-ci-stackrox-stackrox-master-gke-operator-e2e-tests/1706936037810376704) does not look good:
```
    logger.go:42: 08:23:10 | basic-sc/5-central-cr | [pod/scanner-db-84765f58fc-tcp6j/init-db] ALTER TABLE
    logger.go:42: 08:23:10 | basic-sc/5-central-cr | [pod/scanner-db-84765f58fc-tcp6j/init-db] ALTER TABLE
    logger.go:42: 08:23:10 | basic-sc/5-central-cr | [pod/scanner-db-84765f58fc-tcp6j/init-db] ALTER TABLE
    logger.go:42: 08:23:10 | basic-sc/5-central-cr | [pod/scanner-db-84765f58fc-tcp6j/init-db] 2023-09-27 08:19:15.032 UTC [57] ERROR:  canceling autovacuum task
    logger.go:42: 08:23:10 | basic-sc/5-central-cr | [pod/scanner-db-84765f58fc-tcp6j/init-db] 2023-09-27 08:19:15.032 UTC [57] CONTEXT:  automatic analyze of table "postgres.public.vuln_v2"
    logger.go:42: 08:23:10 | basic-sc/5-central-cr | Error from server (BadRequest): container "db" in pod "scanner-db-84765f58fc-tcp6j" is waiting to start: PodInitializing
```

@stackrox/scanner what could cause this?

Same error in both test cases. Interestingly the OCP-based tests passed on first attempt

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml [hunk 1]
```diff
@@ -153,6 +157,10 @@ spec:
           mountPath: /run/secrets/stackrox.io/db-password
         - name: central-external-db-volume
           mountPath: /etc/ext-db
+        {{- if ._rox.central.encryptNotifierSecrets }}
```
Comment (johannes94) [inline image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml:160]: Is there a reason why you used the name  `NotifierSecretsEncryptionEnabled` in CR vs `encryptNotifierSecrets` in the helm chart. Would be simpler to search for and keep changes in sync if both are named the same.

**Review by johannes94 — CHANGES_REQUESTED** [review]
Looks good, only a minor ask about naming of the new value.

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml (hunk 1 above)
Comment (vladbologa) [reply image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml:160]: I was trying to follow the conventions in the surrounding code, but you're right, they should be called the same.

Comment (vladbologa) [reply image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml:160]: Done.

**Review by johannes94 — APPROVED** [review]

File: operator/apis/platform/v1alpha1/central_types.go [hunk 2]
```diff
@@ -135,6 +135,11 @@ type CentralComponentSpec struct {
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=8,displayName="Declarative Configuration",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:hidden"}
 	DeclarativeConfiguration *DeclarativeConfiguration `json:"declarativeConfiguration,omitempty"`
 
+	// Enables the encryption of notifier secrets stored in the Central DB. An encryption key must be
+	// provided in a secret called `central-encryption-key` in the Central namespace.
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=9,displayName="Encrypt Notifier Secrets",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:hidden"}
+	EncryptNotifierSecrets *bool `json:"encryptNotifierSecrets,omitempty"`
```
Comment (vladbologa) [inline operator/apis/platform/v1alpha1/central_types.go:141]: @johannes94 Should this be a struct with a single `Enabled` field, to support future extensions?

E.g. I'm thinking that in the future the Operator might have other options such as `AutoGenerateEncryptionKey`.

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml [hunk 3]
```diff
@@ -153,6 +159,12 @@ spec:
           mountPath: /run/secrets/stackrox.io/db-password
         - name: central-external-db-volume
           mountPath: /etc/ext-db
+        {{- if ._rox.central.notifierSecretsEncryption }}
```
Comment (johannes94) [inline image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml:162]: I'm not sure if we need both if's here or if only `{{- if ._rox.central.notifierSecretsEncryption.enabled }}` is enough.

Did you have problems with CI, your local tests when we didn't have both?

File: operator/apis/platform/v1alpha1/central_types.go [hunk 4]
```diff
@@ -187,6 +199,15 @@ type DeclarativeConfiguration struct {
 	Secrets []LocalSecretReference `json:"secrets,omitempty"`
 }
 
+// NotifierSecretsEncryption defines settings for encrypting notifier secrets in the Central DB.
+type NotifierSecretsEncryption struct {
+	// Enables the encryption of notifier secrets stored in the Central DB. An encryption key must be
+	// provided in a secret called `central-encryption-key` in the Central namespace.
```
Comment (johannes94) [inline operator/apis/platform/v1alpha1/central_types.go:205]: Not sure if the entire filepath will be configurable or has to be correct at the secret level. But for now it looks like they assume a specific filepath, so we could also include that the data field in the secret containing the key should be called `encryptionKey`

**Review by johannes94 — CHANGES_REQUESTED** [review]

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml (hunk 3 above)
Comment (vladbologa) [reply image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml:162]: I was getting many CI failures when _rox.central.notifierSecretsEncryption doesn't exist.

File: operator/apis/platform/v1alpha1/central_types.go (hunk 4 above)
Comment (vladbologa) [reply operator/apis/platform/v1alpha1/central_types.go]: Done

**Review by johannes94 — APPROVED** [review]

**Pull request merged**